uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/addon/EventFilter.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.testing.mqtt.client.control.api.addon;
+
+import com.aws.greengrass.testing.mqtt.client.control.api.ConnectionControl;
+import lombok.Getter;
+import lombok.NonNull;
+
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Filter of events in event storage.
+ */
+@Getter
+public final class EventFilter {
```

Comment:
I still really want to greatly simplify this.

What I'd like is for the agent to emit all events to the UAT - CONNACK, disconnect, message received. We will then handle these events within the UAT itself; not in the agent. The agent should only forward the events and have all the logic inside the UAT.

Reply:
All businesses  logic are in UAT.
EventStorage is a utility to store events received by control currently only received MQTT messages.
In future we can extend for other types of events including unsolicited disconnections and so one.

EventStorage is required in any case when we expecting more that one event i.e. MQTT messages.
EventFilter just a filter in EventStorage used to skip not related events during waiting.
That code developed separately from UAT part because it is a addon for Control and can be developed outside UAT in parallel with building blocks. That utility move complexity from UAT to other place provides clean border between part via API and allow create unit test before UAT part has been developed.

It is really required if we are expecting more that one MQTT message during the time we waiting for broker's delivery.
I have no idea how this can be simplified by moving directly to UAT without implementing a lot of custom error prone cases.


CONNACK, SUBACK, UNSUBACK and so one are passed from agent to control and provided to users of control's API as before.

Reply:
> That code developed separately from UAT part because it is a addon for Control and can be developed outside UAT in parallel with building blocks.

The purpose of all of this code is to support the UAT, we aren't trying to solve abstract problems. We have very concrete problems that need to be implemented.

The UAT should be receiving these events and handling them how it chooses. I do not want an additional unnecessarily complex storage and filtering layer and API.

Reply:
That code works and solve concrete problems. I have no idea how I can simplify that, sorry.

Reply:
Simplify by deleting this code.

What I am suggesting is that all events should be forwarded to the UAT. The UAT can then choose how it wants to handle those events. Storing them in a list, executing callback when an event happens, whatever we want to do. I want the logic to be in the UAT itself, not here.

Reply:
We just move that code to uat, by removing class borders and drop ability to use unit test.
It will  produce solution worst in reading, testing and finally quality.
And nothing will be added.

Reply:
I have several concerns here:
1. Technical. The above example of code functionally not the same as implemented in EventStorageImpl and EventFilter
 a. That code misses filtering by clientId/thingName or connection. That means in "message {string} received on {string} " clientDeviceId is unused. That building block will not be suitable when we have more than a single device. But MQTT5 features like group subscription require that. In other words that implementation is not scalable. It resulting in bigger code in feature and error prone misuse of that building block.
 b.  I have not found documentation on eventuallyEval() but from the name expecting code will check the list after the whole timeout, even when a message is already in the list or received early. It will slow down the tests. Not so critical but disadvantageous.
2. Organizational.
 a. It is hard to do coding on the same file with several programmers, as well as implement receive when no connection or publish is yet implemented. To provide the ability to work in parallel I develop from bottom and other devs from top of the project. This way we must define interfaces as a midway road milestone where we should meet. When other developers did one part of the job I was able to implement functionally isolated classes and test it. Just because we have no other chance to test it before and do not waste time when all untested code dropped to a single pool.
 b. Two developers always implement the same requirement in different ways. It is usual. It depends on their experience and strong and weak  sides. For example if you have good knowledge of modern steaming API lambdas and FI  but I'm not. I prefer the procedural approach because my preferred programming language is still C. As I said it requires time to learn these things but we have a tight schedule.

Don't worry about the number of lines. Complexity of a project depends not on the number of lines but on good (or bad) separation to parts like classes and other black boxes. Number of lines does not affect the cost of the project. But number of lines it mirrors the nature of more complex and flexible architecture with multiple clients and control. Useless to reproduce your already done testing tool, we need build something better and that way defined in project goals and requirements.

Reply:
1.a. of course I just gave a simple example which can be extended with a single line of code to filter for anything else.

1. b. https://github.com/grantwest/eventually-matchers. This does not wait for the timeout before checking, that would be wasteful as you said. It will poll the method on a default or specified interval until the condition succeeds or times out.